### PR TITLE
Enable Outlines with JSON Sub-Schema References

### LINKED ISSUE #15035: [Feature]: Add support for reusable subschemas in tool requests (PydanticAI)
### 🚀 The feature, motivation and pitch

Currently PydanticAI clients leverage tools for structured response mapping. Consider the following ``tools`` definition in the request:

```json
[
    {
        "type": "function",
        "function": {
            "name": "final_result",
            "description": "The final response which ends this conversation",
            "parameters": {
                "$defs": {
                    "Chapter": {
                        "properties": {
                            "chapter_name": {
                                "description": "Name the chapter",
                                "title": "Chapter Name",
                                "type": "string"
                            },
                            "content": {
                                "description": "Content of the chapter",
                                "title": "Content",
                                "type": "string"
                            }
                        },
                        "required": [
                            "chapter_name",
                            "content"
                        ],
                        "title": "Chapter",
                        "type": "object"
                    }
                },
                "properties": {
                    "title": {
                        "description": "Title of the story",
                        "title": "Title",
                        "type": "string"
                    },
                    "summary": {
                        "description": "Short summary of the story",
                        "title": "Summary",
                        "type": "string"
                    },
                    "chapters": {
                        "description": "List of chapters",
                        "items": {
                            "$ref": "#/$defs/Chapter"
                        },
                        "title": "Chapters",
                        "type": "array"
                    }
                },
                "required": [
                    "title",
                    "summary",
                    "chapters"
                ],
                "title": "Story",
                "type": "object"
            }
        }
    }
]
```

Here, ``parameters`` contains the reusable subschema ``Chapter`` passed under ``"$defs"``. This is a valid JSON schema as rendered by a Pydantic ``BaseModel``, however results in a HTTP 400 error in vLLM. 

### Alternatives

For PydanticAI clients there are a few options available:

* Don't use response schemas with nested ``BaseModels``
* Update PydanticAI such that subschemas are de-normalized before calling the vLLM completions endpoint.

## Code to Reproduce

The PydanticAI agent example below can reproduce the issue. Note that it needs to be run against [PR13483](https://github.com/vllm-project/vllm/pull/13483) to work properly with PydanticAI:

```python
class Chapter(BaseModel):
    chapter_name: str = Field(..., description="Name the chapter")
    content: str = Field(..., description="Content of the chapter")

class Story(BaseModel):
    title: str = Field(..., description="Title of the story")
    summary: str = Field(..., description="Short summary of the story")
    chapters: List[Chapter] = Field(..., description="List of chapters")

# Create a PydanticAI agent
agent = Agent(
    name="test_tools3_agent",
    model=llm,
    system_prompt="You are a creative novelist and helpful assistant.",
)

# Fails with error described.
result = await agent.run(
    "Generate a short story about cats.",
    result_type=Story,
)

LOG.info("Results: %s", result.data)
```

If I change ``chapters: List[Chapter]`` to ``chapters: List[str]`` the example runs perfectly. since there is no subschema passed.

## vLLM Startup Command

Note that I am running the version of vLLM from [PR13483](https://github.com/vllm-project/vllm/pull/13483). Basically this PR adds support for ``tool_choice=required`` which is needed for PydanticAI. This is my vLLM run command:

```shell
docker run -it -d \
    --name=eleanor-vLLM \
    --restart=unless-stopped \
    --shm-size=15g \
    --ulimit memlock=-1 \
    --ipc=host \
    --entrypoint=python3 \
    --gpus="device=0,1,2,3" \
    --publish=7800:8000 \
    --volume=/models:/models:ro \
    --health-cmd=timeout 5 bash -c 'cat < /dev/null > /dev/tcp/localhost/8000' \
    --health-start-period=240s \
    --health-interval=15s \
    --health-timeout=8s \
    --health-retries=3 \
    --env=OMP_NUM_THREADS=1 \
    harbor.k8s.wm.k8slab/eleanor-ai/vllm-openai:tool-req-patch \
        -m vllm.entrypoints.openai.api_server \
        --model /models/Llama-3.3-70B-Instruct \
        --served-model-name Llama-3.3-70B-Instruct \
        --response-role auto \
        --load-format safetensors \
        --tokenizer-mode auto \
        --enable-chunked-prefill=True \
        --max-num-batched-tokens=4096 \
        --dtype bfloat16 \
        --kv-cache-dtype auto \
        --gpu-memory-utilization 0.90 \
        --enable-auto-tool-choice \
        --tool-call-parser llama3_json \
        --enable-prefix-caching \
        --device=cuda \
        --task=generate \
        --scheduler-delay-factor=0.25 \
        --uvicorn-log-level=debug \
        --distributed-executor-backend=mp \
        --max-logprobs=100 \
        --enable-prompt-tokens-details \
        --generation-config=auto \
        --override-generation-config={"logprobs": 1} \
        --guided-decoding-backend=outlines \
        --disable_custom_all_reduce \
        --max-model-len 65535 \
        --tensor-parallel-size 4 \
        --port 8000 \
        --host 0.0.0.0
```

## vLLM Logs

Request:

```text
'<|begin_of_text|><|start_header_id|>system<|end_header_id|>\n\nEnvironment: ipython\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\nYou are a creative novelist and helpful assistant.<|eot_id|><|start_header_id|>user<|end_header_id|>\n\nGiven the following functions, please respond with a JSON for a function call with its proper arguments that best answers the given prompt.\n\nRespond in the format {"name": function name, "parameters": dictionary of argument name and its value}.Do not use variables.\n\n{\n    "type": "function",\n    "function": {\n        "name": "final_result",\n        "description": "The final response which ends this conversation",\n        "parameters": {\n            "$defs": {\n                "Chapter": {\n                    "properties": {\n                        "chapter_name": {\n                            "description": "Name the chapter",\n                            "title": "Chapter Name",\n                            "type": "string"\n                        },\n                        "content": {\n                            "description": "Content of the chapter",\n                            "title": "Content",\n                            "type": "string"\n                        }\n                    },\n                    "required": [\n                        "chapter_name",\n                        "content"\n                    ],\n                    "title": "Chapter",\n                    "type": "object"\n                }\n            },\n            "properties": {\n                "title": {\n                    "description": "Title of the story",\n                    "title": "Title",\n                    "type": "string"\n                },\n                "summary": {\n                    "description": "Short summary of the story",\n                    "title": "Summary",\n                    "type": "string"\n                },\n                "chapters": {\n                    "description": "List of chapters",\n                    "items": {\n                        "$ref": "#/$defs/Chapter"\n                    },\n                    "title": "Chapters",\n                    "type": "array"\n                }\n            },\n            "required": [\n                "title",\n                "summary",\n                "chapters"\n            ],\n            "title": "Story",\n            "type": "object"\n        }\n    }\n}\n\nGenerate a short story about cats.<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n', params: SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=1.0, top_p=1.0, top_k=-1, min_p=0.0, seed=None, stop=[], stop_token_ids=[], bad_words=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=12000, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None, guided_decoding=GuidedDecodingParams(json={'type': 'array', 'minItems': 1, 'items': {'type': 'object', 'anyOf': [{'properties': {'name': {'type': 'string', 'enum': ['final_result']}, 'parameters': {'$defs': {'Chapter': {'properties': {'chapter_name': {'description': 'Name the chapter', 'title': 'Chapter Name', 'type': 'string'}, 'content': {'description': 'Content of the chapter', 'title': 'Content', 'type': 'string'}}, 'required': ['chapter_name', 'content'], 'title': 'Chapter', 'type': 'object'}}, 'properties': {'title': {'description': 'Title of the story', 'title': 'Title', 'type': 'string'}, 'summary': {'description': 'Short summary of the story', 'title': 'Summary', 'type': 'string'}, 'chapters': {'description': 'List of chapters', 'items': {'$ref': '#/$defs/Chapter'}, 'title': 'Chapters', 'type': 'array'}}, 'required': ['title', 'summary', 'chapters'], 'title': 'Story', 'type': 'object'}}, 'required': ['name', 'parameters']}]}}, regex=None, choice=None, grammar=None, json_object=None, backend=None, whitespace_pattern=None), extra_args=None), prompt_token_ids: None, lora_request: None, prompt_adapter_request: None.
INFO 03-18 12:20:23 [async_llm_engine.py:549] Building guided decoding logits processor. guided_decoding: GuidedDecodingParams(json={'type': 'array', 'minItems': 1, 'items': {'type': 'object', 'anyOf': [{'properties': {'name': {'type': 'string', 'enum': ['final_result']}, 'parameters': {'$defs': {'Chapter': {'properties': {'chapter_name': {'description': 'Name the chapter', 'title': 'Chapter Name', 'type': 'string'}, 'content': {'description': 'Content of the chapter', 'title': 'Content', 'type': 'string'}}, 'required': ['chapter_name', 'content'], 'title': 'Chapter', 'type': 'object'}}, 'properties': {'title': {'description': 'Title of the story', 'title': 'Title', 'type': 'string'}, 'summary': {'description': 'Short summary of the story', 'title': 'Summary', 'type': 'string'}, 'chapters': {'description': 'List of chapters', 'items': {'$ref': '#/$defs/Chapter'}, 'title': 'Chapters', 'type': 'array'}}, 'required': ['title', 'summary', 'chapters'], 'title': 'Story', 'type': 'object'}}, 'required': ['name', 'parameters']}]}}, regex=None, choice=None, grammar=None, json_object=None, backend=None, whitespace_pattern=None)
INFO:     172.17.0.1:21179 - "POST /v1/chat/completions HTTP/1.1" 400 Bad Request
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by theobjectivedad**
## Background

This update applies (optional) pre-processing logic to the Outlines guided decoding implementation. During the guided decoding process, vLLM invokes ``build_regex_from_schema`` on the provided JSON schema string to return a regex that corresponds to the schema. During generation, this regex is used to set logits to ``-inf`` that do not correspond to the requested schema.

This issue was discovered testing [PydanticAI](https://ai.pydantic.dev/) which relied on invoking tool functions to format output. While the PydanticAI methodology works as expected with simple outputs, Pydantic generates JSON subschemas (think "$defs" and "$refs") for nested ``BaseModel``s and raises an error in vLLM.

Note that for full support, PydanticAI need ``tool_use=required``, see PR #13483.

## Triggering

To trigger the issue and demonstrate the fix, the following request can be sent:

```shell
curl --location 'https://YOUR_URL_GOES_HERE/v1/chat/completions' \
--header 'Content-Type: application/json' \
--data '{
    "messages": [
        {
            "role": "user",
            "content": "Generate a short outline on about cars."
        }
    ],
    "model": "QwQ-32B",
    "n": 1,
    "presence_penalty": 0.5,
    "stream": false,
    "tool_choice": "auto",
    "tools": [
        {
            "type": "function",
            "function": {
                "name": "final_result",
                "description": "Contains a simple outline of a topic",
                "parameters": {
                    "$defs": {
                        "HeadingDetail": {
                            "description": "An outline detail record",
                            "properties": {
                                "content": {
                                    "description": "Content of the heading",
                                    "title": "Content",
                                    "type": "string"
                                }
                            },
                            "required": [
                                "content"
                            ],
                            "title": "HeadingDetail",
                            "type": "object"
                        }
                    },
                    "properties": {
                        "heading": {
                            "description": "Heading details",
                            "items": {
                                "$ref": "#/$defs/HeadingDetail"
                            },
                            "title": "Heading",
                            "type": "array"
                        }
                    },
                    "required": [
                        "heading"
                    ],
                    "title": "Outline",
                    "type": "object"
                }
            }
        }
    ]
}'
```

To replicate, I used the following vLLM run command:

```shell
python -m vllm.entrypoints.openai.api_server \
		--model=/models/QwQ-32B \
		--served-model-name=QwQ-32B \
		--gpu-memory-utilization=0.90 \
		--override-generation-config='{"temperature": 0.6, "top_p": 0.95, "top_k": 30, "add_generation_prompt": true }' \
		--rope-scaling='{"factor": 3.0, "original_max_position_embeddings": 40960, "rope_type": "yarn"}' \
		--enable-auto-tool-choice \
		--tool-call-parser=hermes \
		--guided-decoding-backend=outlines \
		--enable-prefix-caching \
		--enable-prompt-tokens-details \
		--generation-config=auto \
		--device=cuda \
		--disable-custom-all-reduce \
		--distributed-executor-backend=mp \
		--dtype=auto \
		--host=0.0.0.0 \
		--port=7800 \
		--kv-cache-dtype=auto \
		--load-format=safetensors \
		--max-logprobs=100 \
		--max-num-batched-tokens=4096 \
		--task=generate \
		--tensor-parallel-size=4 \
		--uvicorn-log-level=debug
```

Fix #15035

# Bugfixes

* Adding small bugfix that allows ``tool_choice=required`` to work with reasoning models.

### CODE REVIEW

**Comment by webcoderz** [comment]
Can we get a review here @russellb , @mgoin or anybody? This fixes nested pydantic models in structured outputs

**Comment by RWayne93** [comment]
Ran into this issue today and crashed my production vllm pod. Is this almost ready to be merged? was about to work on a fix myself before seeing this PR.

**Comment by webcoderz** [comment]
> ```shell
> 		--guided-decoding-backend=outlines \
> ```

WIP, looks like maintainers have added it to a list of PRs to review for structured output support

**Comment by theobjectivedad** [comment]
Good morning @mgoin / @russellb , this will also address tool calling issues with FastAPI/vLLM as well; basically any client framework that uses guided decoding to generate structured outputs would benefit. Would you have a moment to take a look at this PR?

**Comment by russellb** [comment]
We are trying to avoid any major changes to the V0 engine. #15975 is adding V1 support. Perhaps you could look at that? or enabling what you're looking for using one of the other backends already included in V1?

**Comment by RWayne93** [comment]
its unclear to me will #15975 fix this issue? Pydantic Ai has been my preferred agent framework but haven't been able to use it to its full potential with vllm because of this issue.

**Comment by theobjectivedad** [comment]
Good morning @russellb, thank you for looking into this. I do agree there this is a new capability in V0/Outlines guided decoding. However I do argue that the scope of this PR is (a) isolated to providing behavior that folks would/should normally expect and (b) becoming increasingly relevant for teams who rely on guided decoding to produce structured outputs (PydanticAI, FastAgent, etc.).

I'm happy to wait to see if #15975 addresses the problem however, I also request that you reconsider this PR as well for folks who cannot immediately switch to V1.

**Comment by theobjectivedad** [comment]
Hello again @russellb, sorry to nag but do you have any additional thoughts in this? If you are aligned I can rebase.

**Comment by TensorTemplar** [comment]
this should a billion percent remain open

**Comment by webcoderz** [comment]
agreed
